### ISSUE TITLE
Bazel test doesn't work on Windows x64 machine.

### ISSUE THREAD

**Issue by steed924** [TraceMachina/nativelink#485]
I've tried Bazel test on Windows x64 machine.
I ran the server with `cargo `command and it started running correctly.
`cargo run --bin cas -- ./nativelink-config/examples/basic_cas.json`

But `bazel test` command fails with this error message.
Is it a known issue for us? I think we didn't add `Windows x64` configuration for the local execution build file.

![Capture](https://github.com/TraceMachina/nativelink/assets/93139780/e4fd9600-7b11-40d1-9bb7-60c913eca1f1)

**Comment by aaronmondal** [comment]
Oh I think this is caused by #471 and just got lost in the logs. We're importing `rules_cc` in the `WORKSPACE` file now and registering additional toolchains. I'm surprised that this is influencing non-`--config=lre` builds.

**Comment by steed924** [comment]
Hello, @aaronmondal .
Thanks for considering this issue.
I'd like to work on Windows x64 machine from now since none of us are using Windows for development.
Previously, when I worked on Ubuntu, those errors didn't arise and it seems like they are clearly platform-specific error.
I'd like to work on this issue with you. ❤️

**Comment by steed924** [comment]
Plus, I found out that running `rustfmt` via `bazel` wasn't working as well on Windows x64 machine.
I will be working on this issue as well to set up my development environment on Windows x64 machine.
![Capture](https://github.com/TraceMachina/nativelink/assets/93139780/93eee28b-1474-4db0-a365-841c0bc14d68)

**Comment by aaronmondal** [comment]
> I'd like to work on this issue with you. ❤️
I'd love to as well :heart: 

Yesterday we ran into the same thing on MacOS which is currently not enabled in CI but was kinda working for ~4 days before #471 broke it :sweat_smile: 

For more context to what I think is happening:

#471 Introduced `rules_cc` and some autogenerated nix-backed C++ toolchains that require `--incompatible_enable_cc_toolchain_resolution` to be set. I assumed that this wouldn't affect anything that was not using the `--config=lre`. However, these lines here seem to "override" the default resolution mechanism even in the default config:

https://github.com/TraceMachina/nativelink/blob/96302cbeab6c59966d3dfd3b99fa0933752d1018/WORKSPACE.bazel#L12-L19

On linux this isn't noticeable as the toolchain suite that triggers this is valid on linux:

https://github.com/TraceMachina/nativelink/blob/96302cbeab6c59966d3dfd3b99fa0933752d1018/local-remote-execution/generated/cc/BUILD#L47-L58

On non-nix this toolchain doesn't get selected though and so the native-Bazel CI workflows keep working as expected. So far this is intended behavior.

The issue seems to be that the above `cc_toolchain_suite` is not valid on Windows and MacOS. Usually I'd expect this to not be an issue because I'd assume that Bazel just "skips" this suite entirely. Also, the platform/toolchain that is associated with this toolchain_suite is explicitly marked to only be compatible with x86_64-linux:

https://github.com/TraceMachina/nativelink/blob/96302cbeab6c59966d3dfd3b99fa0933752d1018/local-remote-execution/generated/config/BUILD#L18-L47

---

There was an issue in Bazel that I can't find anymore which suggested that this might be fixed in bazel 7.0.0. I think it might be worth investigating that. Otherwise it might work if we register additional toolchains for Windows and MacOS which wrap regular toolchain autodetection.

Also one note for testing: The workspace is read in serial. Moving things like the `register_*` and `rules_cc`  stuff around *does* influence how this issue manifests, but so far I haven't found an import order that satisfies all usecases :laughing:

**Comment by steed924** [comment]
Thanks for your reply, @aaronmondal.
Just fyi, I did some testing on my machine after I rollback the Local Remote Execution changes and tried to see what happens.
This time, `bazel test` didn't raise error but while execution, it failed with following messages.
I'd like to know if we've done any Windows testing before. Thanks.

![Capture](https://github.com/TraceMachina/nativelink/assets/93139780/4bd8b10b-be39-4aeb-abae-55dbe35c34c7)

**Comment by aaronmondal** [comment]
@steed924 Yes, the previous runs on main for the native Windows Bazel build worked are here: https://github.com/TraceMachina/nativelink/actions/workflows/native-bazel.yaml?query=branch%3Amain

However, we did not run integration tests on Windows. Here's what's happening:

- Bazel rules internally construct `action`s which are more or less shell commands that create a sandbox with some environment variables and an action command.
- When you use remote execution you construct the "action string" locally and then Nativelink forwards that to some remote worker.
- In your case that remote worker is your local machine.
- The command failing means that the build environment is most likely not hermetic, or at least that the local environment and the remote execution environment don't perfectly align. Locally you have the files available, and so the action string contains these strings to local paths. But remotely these files don't exist and so the fails when invoked in this sandbox.
- This issue was kind of hard to see because the current integration tests (on linux) brute-force mount local directories into the remote environment. That is, the tests don't actually create a "true" remote execution environment but a hybrid that has visibility into the local filesystem.
- #471 and #479 address this by creating truly encapsulated remote execution environments where the remote execution container is completely airgapped from the local environment. This also exposed some issues in rules_rust (#477) and building nlink in an airgapped remote execution setup is currently not possible. The LRE setup only works for C++.

The only environment that accurately constructs a remote execution setup is the K8s example. That doesn't work on Windows and we can't make that exact setup compatible with Windows because Nix only works in WSL2 but not on native Windows.

However, what we *can* do is distribute a C++ toolchain for windows as part of the Bazel build which overrides any local compilers. This way we get this "airgapped" setup on windows as well. When we ship the toolchain we're no longer susceptible to differing host toolchain configurations and I think that's the way to fix this issue.

FYI the workaround which is used in the (linux) production deployments at the moment is to run nativelink in container images that "just happen" to have all relevant dependencies at the right places. For instance by using a `FROM ubuntu` with an `apt install clang` image you'd get "some" compiler at `/usr/bin/clang` that just happens to work. That compiler isn't binary-identical to the one on your hypothetical local Ubuntu machine, but the paths are the same and so things still work.

Note that this is actually quite bad because it means that even in the cases that currently work the same action command will create different artifacts on your local machine and a remote machine. In terms of reproducibility/hermeticity I'd consider this a "critical silent bug" for all existing deployments except the LRE setup.

**Comment by aaronmondal** [comment]
Oh and actually regarding your initial question, remote execution on Windows was not tested and I don't think it ever worked :sweat_smile: We'll have to implement a windows toolchain for remote execution. From some initial googling I couldn't find any existing remote-exec compatible Windows toolchains.

**Comment by p00f** [comment]
@aaronmondal 
Hi, I have a similar error (note that it fails to find the params file):

```
 Error { code: NotFound, messages: ["The system
cannot find the path specified. (os error 3)", "Could not execute command [\"external/MINGW_TOOLCHAIN/bin/g++.exe\", \"@bazel-out/windows_platform-fastbuild/bin/xxxxx/_objs/TARGET_NAME/OBJECT_NAME.o.params\"]"] }
```
> The command failing means that the build environment is most likely not hermetic, or at least that the local environment and the remote execution environment don't perfectly align. Locally you have the files available, and so the action string contains these strings to local paths. But remotely these files don't exist and so the fails when invoked in this sandbox.

I don't think so, at least in my case - there _is_ such a file in the directory specified in the configs `workers.local.work_directory`/SOME_HASH/work, except without the `@` in the beginning.

Also, the hash in `workers.local.work_directory`/SOME_HASH/work changes periodically
